vendor/github.com/hashicorp/vault/api/renewer.go

```diff
@@ -269,11 +269,7 @@ func (r *Renewer) renewLease() error {
 		// Grab the lease duration and sleep duration
 		leaseDuration := time.Duration(renewal.LeaseDuration) * time.Second
 		sleepDuration := r.sleepDuration(leaseDuration)
-
-		// If we are within grace, return now.
-		if leaseDuration <= r.grace || sleepDuration <= r.grace {
-			return nil
-		}
```

Comment:
I removed this entire if clause, It would cause the renew method to return instantly because it was checking against the vault grace period. With this removed, it correctly sleeps for sleepDuration before returning. The return is interpreted as a write to the DoneCh which causes the loop in vault_read.go to go renew the token again. After I removed this and tested the same template, it correctly only renews according to leaseDuration.